File: 2021/day03_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    parsed = [parse(line) for line in content if line]

    return to_int(filter_most_common(parsed, 0)) * to_int(
        filter_least_common(parsed, 0)
    )


def filter_most_common(content: list[list[int]], at: int):
    if len(content) == 1:
        return content[0]
    count = count_all(content)
    (c0, c1) = count[at]
    return filter_most_common(
        [line for line in content if line[at] == (0 if c0 > c1 else 1)],
        at + 1,
    )


def filter_least_common(content: list[list[int]], at: int):
    if len(content) == 1:
        return content[0]
    count = count_all(content)
    (c0, c1) = count[at]
    return filter_least_common(
        [line for line in content if line[at] == (0 if c0 <= c1 else 1)],
        at + 1,
    )


def to_int(binary: list[int]):
    return sum(bit * 2**index for index, bit in enumerate(reversed(binary)))


def parse(line: str):
    return [int(x) for x in line.strip()]


def init(length: int):
    return [(0, 0)] * length


def count_all(content: list[list[int]]):
    acc = init(len(content[0]))
    for line in content:
        acc = count_line(acc, line)
    return acc


def count_line(acc: list[tuple[int, int]], line: list[int]):
    return [
        (c0 + (1 if bit == 0 else 0), c1 + (1 if bit == 1 else 0))
        for ((c0, c1), bit) in zip(acc, line)
    ]


def main() -> None:
    with open("day03.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2021/day03-2.ts
type Count = { c0: number; c1: number };

function solution(content: string) {
  const parsed = content
    .split("\n")
    .map((line) => line.trim().split("").map(Number));

  return (
    toInt(filterMostCommon(parsed, 0)) * toInt(filterLeastCommon(parsed, 0))
  );
}

function filterMostCommon(content: number[][], at: number) {
  if (content.length === 1) return content[0];
  const count = countAll(content);
  const { c0, c1 } = count[at];
  return filterMostCommon(
    content.filter((line) => line[at] === (c0 > c1 ? 0 : 1)),
    at + 1
  );
}

function filterLeastCommon(content: number[][], at: number) {
  if (content.length === 1) return content[0];
  const count = countAll(content);
  const { c0, c1 } = count[at];
  return filterLeastCommon(
    content.filter((line) => line[at] === (c0 <= c1 ? 0 : 1)),
    at + 1
  );
}

function toInt(binary: number[]) {
  return binary
    .reverse()
    .reduce((acc, cur, index) => acc + cur * Math.pow(2, index), 0);
}

function countAll(content: number[][]) {
  return content.reduce(
    (acc, cur) =>
      acc.map(({ c0, c1 }, index) => ({
        c0: c0 + (cur[index] === 0 ? 1 : 0),
        c1: c1 + (cur[index] === 1 ? 1 : 0),
      })),
    Array.from({ length: content[0].length }).map(() => ({
      c0: 0,
      c1: 0,
    })) as Count[]
  );
}

Deno.readTextFile("day03.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
